app/gui/src/controller/graph/widget/configuration.rs

```diff
@@ -51,34 +52,88 @@ pub fn deserialize_widget_definitions(
 
 /// Convert a widget definition from the engine response into a IDE internal widget configuration
 /// struct. See [`widget::Configuration`] for more information.
-fn to_configuration(resp: response::WidgetDefinition) -> widget::Configuration {
-    widget::Configuration { display: resp.display, kind: to_kind(resp.inner), has_port: true }
+fn to_configuration(
+    resp: response::WidgetDefinition,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> widget::Configuration {
+    widget::Configuration {
+        display:  resp.display,
+        kind:     to_kind(resp.inner, db, parser),
+        has_port: true,
+    }
 }
 
-fn to_kind(inner: response::WidgetKindDefinition) -> widget::DynConfig {
+fn to_kind(
+    inner: response::WidgetKindDefinition,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> widget::DynConfig {
     match inner {
-        response::WidgetKindDefinition::SingleChoice { label, values } =>
+        response::WidgetKindDefinition::SingleChoice { label, values } => {
+            let (entries, arguments) = to_entries_and_arguments(values, db, parser);
             widget::single_choice::Config {
-                label:   label.map(Into::into),
-                entries: Rc::new(to_entries(values)),
+                label: label.map(Into::into),
+                entries: Rc::new(entries),
+                arguments,
             }
-            .into(),
+            .into()
+        }
         response::WidgetKindDefinition::ListEditor { item_widget, item_default } =>
             widget::list_editor::Config {
-                item_widget:  Some(Rc::new(to_configuration(*item_widget))),
+                item_widget:  Some(Rc::new(to_configuration(*item_widget, db, parser))),
                 item_default: ImString::from(item_default).into(),
             }
             .into(),
         _ => widget::label::Config::default().into(),
     }
 }
 
-fn to_entries(choices: Vec<response::Choice>) -> Vec<widget::Entry> {
-    choices.into_iter().map(to_entry).collect()
+fn to_entries_and_arguments(
+    choices: Vec<response::Choice>,
+    db: &SuggestionDatabase,
+    parser: &parser::Parser,
+) -> (Vec<widget::Entry>, Vec<(usize, ImString, widget::Configuration)>) {
```

Comment:
I didn't want to put a `Vec<Configuration>` into the widget entry, because it would require us to keep all widget configurations as `Hash, PartialEq, Eq`. That would mean no floats allowed. This is definitely a no-go, as we will very soon have number picker with all kinds of limits or step values. This is the primary reason for keeping it a separate list, though better perf is also a bonus (no need to clone that all the time into the dropdown view itself).

As for naming, I guess renaming `Entry` to `Choice` could work better, as then you can easier mentally link that to "single choice" or potentially "multiple choice" widgets.